keps/sig-api-machinery/2558-publish-version-openapi/README.md

```diff
@@ -0,0 +1,415 @@
+
+### Example
+
+Consider adding a new alpha API field `Width` to the stable API `Frobber`
+in `v1.20`.
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // This field is alpha-level and is only honored by servers that enable the Frobber2D feature.
+  // +optional
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+The field `Width` is controlled by the `Frobber2D` feature gate defined
+in [`kube_features.go`].
+
+```go
+// owner: @you
+// alpha: v1.20
+//
+// Add multiple dimensions to frobbers.
+Frobber2D utilfeature.Feature = "Frobber2D"
+
+var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureSpec{
+  ...
+  Frobber2D: {Default: false, PreRelease: utilfeature.Alpha},
+}
+```
+
+To link the feature gate with the API field, the following comment line
+is added for the `Width` field:
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // +optional
+  // +lifecycle=alpha
+  // +kubernetes:minVersion=v1.20
+  // +featureGate=Frobber2D
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+This will generate an OpenAPI schema with an additional vendor extension field:
+
+```go
+"width": {
+  VendorExtensible: spec.VendorExtensible{
+    Extensions: spec.Extensions{
+      "x-kubernetes-version": map[string]interface{}{
+        "lifecycle":  "alpha",
+        "kubernetes:minVersion":  "v1.20",
+        "featureGate": "Frobber2D",
+        },
+      },
+    },
+  },
+```
+
+This shows that the field is alpha from v1.20 and 
+is controlled by the `Frobber2D` feature gate.
+
+### Constraints and Caveats
+
+#### Constraints
+
+1. The source of truth for versioning information for a feature gate
+will be the comment tags on the respective API fields because
+these comment tags are used for publishing to OpenAPI.
+A linter ensures that comments for feature gates defined in
+[`kube_features.go`] do not diverge from the comment lines on API fields.
+
+2. Additionally, the following constraints are applied and validated for
+comment tags on API fields:
+
+- If an API type contains the `+lifecycle` tag, it must contain
+the `+kubernetes:minVersion` and `+featureGate` tags too.
+- The `+lifecycle` value must be one of `alpha`, `beta`, or `deprecated`.
+- The `+kubernetes:minVersion` value must satisfy the regular expression `^v[1-9][0-9]*\.(0|[1-9][0-9]*)$`.
+- The `+featureGate` value is a valid feature gate as defined in
+[`kube_features.go`].
+
+### Test Plan
+
+The constraints for the comment tags listed above are validated using
+a linter and relevant tests for introducing a new OpenAPI extension
+are added to kube-openapi.
+
+### Graduation Criteria
+
+This feature is currently in the alpha status.
+
+For the feature to move to the beta status, feedback will be collected
+on usage of the `x-kuberentes-field-version` extension in OpenAPI in
```

Comment:
you forgot that one :-)

Reply:
And actually there's a small typo to Kubernetes

Reply:
> And actually there's a small typo to Kubernetes

ah, that feeling when you are trying to rush to get through the KEP by the enhancement freeze :)

Will fix in a follow up :+1: